### DISCUSSION TITLE
Backward compatible ESM support

### DISCUSSION THREAD

**Discussion by nbouvrette** [vercel/next.js#33148] [Ideas]
### Describe the feature you'd like to request

Since Next.js `11.1.3-canary.69`, ESM externals has been turned on by default using the `experimental: { esmExternals: true }` in this [commit](https://github.com/vercel/next.js/pull/29878).

One of the challenges is that the module resolutions between CommonJS and ESM is different. From Node [documentation](https://nodejs.org/api/esm.html):

> The current specifier resolution does not support all default behavior of the CommonJS loader. One of the behavior differences is automatic resolution of file extensions and the ability to import directories that have an index file.

By having this, it means that you could end up with different import paths depending on if you use CommonJS or ESM which will be a really bad developer experience. On top of this, to provide the correct support for Next.js, modules would have to be backward compatible since not all packages support ESM today.

### Describe the solution you'd like

It looks like the way to do would be to support `--experimental-specifier-resolution=node` but from looking at previous issues ([here](https://github.com/vercel/next.js/issues/9607) and [here](https://github.com/vercel/next.js/issues/9487)), it does not seem to work with Next.js (not sure why).

I was not able to get it to work myself, so maybe we could have this as a new experimental option that is supported by Next.js?

### Describe alternatives you've considered

I tried to build a backward compatible package using this [article](https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html) but ended up with clunky paths which degraded the developer experience.

What I mean by clunky paths is:

- For example, if we have an import that points to `next-multilingual/messages` we would now need to use `next-multilingual/messages.js`
- To get `next-multilingual/messages.js` to work we would need to remove the file in `src/messages/index.js` and move it to `src/messages.js` which then prevents us from using directories to organize files which will make the project messier
- Or we could use `next-multilingual/messages/messages.js` to keep the directory but this makes the import even harder to read and less intuitive

I believe the experimental flag would help resolve some of these issues and is worth trying.

**Comment by balazsorban44** [comment]
You can disable that new behavior by `esmExternals: false` at the top-level (not experimental) of the config for backwards compatibility.

**Reply by nbouvrette** [reply]
@balazsorban44 yes I realize that, but by doing this we are not benefiting from ESM? I also presume the long-term strategy is to move to ESM? If that is the case, increasing backward compatibility options should give us a better chance to get there?

One other thing is that when we use `esmExternals: false` there is a warning that we use experimental features which I'm not sure is accurate in this scenario :)

**Reply by nbouvrette** [reply]
And sorry just saw your mention of "top level" - I guess you can ignore my second comment there - but the first one is still valid.

**Comment by nbouvrette** [comment]
> You can disable that new behavior by esmExternals: false at the top-level (not experimental) of the config for backwards compatibility.

@balazsorban44 I was about to make a change on my repo and [I could not find any reference](https://github.com/vercel/next.js/search?q=esmExternals) to a "top-level (not experimental)" `esmExternals`?

I am planning to add an issue because this warning is misleading.

**Reply by nbouvrette** [reply]
For reference, I opened this issue: https://github.com/vercel/next.js/issues/33621
